core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -691,39 +703,62 @@ private void freeSpaceInternal(long sessionId, long availableBytes, BlockStoreLo
-        try {
-          moveResult = moveBlockInternal(Sessions.createInternalSessionId(),
-              blockId, oldLocation, newLocation);
-        } catch (InvalidWorkerStateException e) {
-          // Evictor is not working properly
-          LOG.error("Failed to demote blockId {}, this is temp block", blockId);
-          continue;
-        } catch (BlockAlreadyExistsException e) {
-          continue;
-        } catch (BlockDoesNotExistException e) {
-          LOG.info("Failed to demote blockId {}, it could be already deleted", blockId);
-          continue;
-        }
-        if (moveResult.getSuccess()) {
-          synchronized (mBlockStoreEventListeners) {
-            for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-              listener.onMoveBlockByWorker(sessionId, blockId, moveResult.getSrcLocation(),
-                  newLocation);
+        moveTasks.add(() -> {
+          long blockId = entry.getBlockId();
+          BlockStoreLocation oldLocation = entry.getSrcLocation();
+          BlockStoreLocation newLocation = entry.getDstLocation();
+
+          try {
+            MoveBlockResult moveResult = moveBlockInternal(Sessions.createInternalSessionId(),
+                blockId, oldLocation, newLocation);
+
+            if (moveResult.getSuccess()) {
+              synchronized (mBlockStoreEventListeners) {
+                for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+                  listener.onMoveBlockByWorker(sessionId, blockId, moveResult.getSrcLocation(),
+                      newLocation);
+                }
+              }
+              return true;
             }
+          } catch (InvalidWorkerStateException e) {
+            // Evictor is not working properly
+            LOG.error("Failed to demote blockId {}, this is temp block", blockId);
+          } catch (BlockAlreadyExistsException e) {
+            LOG.info("Failed to demote blockId {}, it could be already moved", blockId);
+          } catch (BlockDoesNotExistException e) {
+            LOG.info("Failed to demote blockId {}, it could be already deleted", blockId);
           }
-        }
+          return false;
+        });
+      }
+      // Execute move orders in thread pool.
+      try {
+        List<Future<Boolean>> moveResults = mEvictionExecutor.invokeAll(moveTasks);
+        long failCount = moveTasks.size() - moveResults.stream().filter(
+            (fut) -> {
+              try {
+                return fut.get();
+              } catch (Exception e) {
+                return false;
+              }
+            }).count();
+        LOG.info("Finished moving:{} blocks to tier: {}. Failure count: {} ", moveTasks.size(),
+            mStorageTierAssoc.getAlias(tierOrdinal), failCount);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException("Interrupted while executing eviction move orders.");
```

Comment:
Should we attempt to cancel or interrupt the tasks running in the eviction executor? Also, is throwing runtime exception here handled properly by the callers?

Reply:
`invokeAll` cancels unfinished tasks if interrupted.

Reply:
Caller was crash-exiting. I declared `InterruptedException` along the call chain to gracefully handle it.

Reply:
How does `InterruptedException` get handled further up the stack? If eviction is interrupted, should we just interpret that as out of space?

Reply:
Not throwing eviction will cause cascading eviction runs to continue until the heartbeat executor notices the interruption. And we'd need to properly store interruption status to heartbeat executor.